diopter/compiler.py

```diff
@@ -359,39 +360,135 @@ def __exit__(
             os.remove(self.path)
 
 
-@dataclass(frozen=True, kw_only=True)
-class CompilationInfo:
-    source_file: Path
-    stdout_stderr_output: str
-
+def temporary_file(contents: str, suffix: str) -> IO[bytes]:
+    ntf = tempfile.NamedTemporaryFile(suffix=suffix)
+    if contents:
+        with open(ntf.name, "w") as f:
+            f.write(contents)
+    return ntf
 
-class CompilationOutputKind(Enum):
-    Object = 0
-    Assembly = 1
-    Exec = 2
-    Unspecified = 3
 
-    def to_flag(self) -> str:
-        match self:
-            case CompilationOutputKind.Object:
-                return "-c"
-            case CompilationOutputKind.Assembly:
-                return "-S"
-            case CompilationOutputKind.Exec | CompilationOutputKind.Unspecified:
-                return ""
+class CompilationOutput(ABC):
+    def __init__(self, filename: Path | None) -> None:
+        if filename:
+            self.filename = filename
+            self.temporary_file = None
+        else:
+            self.temporary_file = tempfile.NamedTemporaryFile(
+                suffix=type(self).suffix(), delete=False
+            )
+            self.temporary_file.close()
+            self.filename = Path(self.temporary_file.name)
 
-
-@dataclass(frozen=True)
-class CompilationOutput:
-    filename: Path
-    kind: CompilationOutputKind
+    def __del__(self) -> None:
+        if self.temporary_file is None:
+            return
+        if Path(self.filename).exists():
+            os.remove(self.filename)
 
     def to_cmd(self) -> str:
-        if self.kind == CompilationOutputKind.Unspecified:
+        if type(self).empty_command():
             return ""
-        if kind_flag := self.kind.to_flag():
-            return kind_flag + " -o " + str(self.filename)
-        return "-o " + str(self.filename)
+        return type(self).flag() + " -o " + str(self.filename)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, CompilationOutput):
+            raise NotImplementedError
+        return (
+            type(self) == type(other)
+            and self.filename == other.filename
+            and self.temporary_file == other.temporary_file
+        )
+
+    @staticmethod
+    @abstractmethod
+    def flag() -> str:
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def suffix() -> str:
+        raise NotImplementedError
+
+    @staticmethod
+    def empty_command() -> bool:
+        return False
```

Comment:
What about making `to_cmd` to a `str | None`?

Pro:
- will not forget to check
Con:
- Has to check non-none-nes even when it would be clear that the command is non-empty. (Might override with str only?)

Maybe adding a bit of documentation is sufficient.

Reply:
You're commenting on an older revision :P

Reply:
This `empty_command` is only relevant for parsing `CompilationSettings` without an explicit output. I'll add this to the docstring